### ISSUE TITLE
bzlmod issues in 0.12.0

### ISSUE THREAD

**Issue by novas0x2a** [theoremlp/rules_multitool#57]
I'm converting our repo to use bzlmod, and I noticed a few things in the process; I'm sorry for the lack of details on some of this, I'm new to bzlmod so I can currently only point out issues, not solutions :|
- [this todo](https://github.com/theoremlp/rules_multitool/blob/2daa11c79ba7ad5a56b9af14acc066f7aa79c933/multitool/extension.bzl#L17) means you can't use `rules_lint` and `rules_multitool` at the same time if you pin any dep using multitool that might also be pinned in `rules_lint` (now or in the future) since it all ends up in the same hub. (I thought this was supposed to be namespaced with bzlmod, but maybe not?)
- In `MODULE.bazel.lock`, the key `@@rules_multitool~//multitool:extension.bzl%multitool` / `general` / `bzlTransitiveDigest` varies depending on the platform you generate it on; in my case, I generated it on linux, then switched to windows and ran with `--lockfile_mode=error` and it broke.

**Comment by mark-thm** [comment]
> [this todo](https://github.com/theoremlp/rules_multitool/blob/2daa11c79ba7ad5a56b9af14acc066f7aa79c933/multitool/extension.bzl#L17) means you can't use rules_lint and rules_multitool at the same time if you pin any dep using multitool that might also be pinned in rules_lint (now or in the future) since it all ends up in the same hub. (I thought this was supposed to be namespaced with bzlmod, but maybe not?)

Not anything to do with bzlmod here -- the TODO you're flagging would allow for some additional flexibility but comes at some cost to usability.

That said, rules_lint by design wants you to be able to override the rules_lint tool versions, which you can do because root module versions for the same tool _name_ will override what's in rules_lint. But, you can have multiple binary versions for the same underlying tool by setting a different tool name for each version.

> In MODULE.bazel.lock, the key @@rules_multitool~//multitool:extension.bzl%multitool / general / bzlTransitiveDigest varies depending on the platform you generate it on; in my case, I generated it on linux, then switched to windows and ran with --lockfile_mode=error and it broke.

I think I'd need more information or at least a minimal reproduction to provide any guidance here. We are long-time users of this ruleset and of bzlmod but do not commit bzlmod lockfiles for several reasons...but I've never had a problem enabling lockfiles and using this ruleset, so maybe something else is at play.

**Comment by novas0x2a** [comment]
I'm working on repros, but having trouble so far, but I'll give you more info and see if that helps. So, re: the rules_lint problem, the context is we're already using multitool and rules_lint, and I'm now migrating us to bzlmod. So I went from this (note: a subset, not the whole file, sorry about that):
```bazel
github_archive(
    name = "aspect_rules_lint",
    commit = "369cdb2725aa75a1de3cc580a051e8dca9b87c30",
    owner = "aspect-build",
    repo_name = "rules_lint",
    sha256 = "55f7492ae2ecdd65bb17925f8eb8b747e1b5b0e3def4ffa6d1cc1243e3f7a03e",
)

github_archive(
    name = "rules_multitool",
    commit = "v0.12.0",
    owner = "theoremlp",
    sha256 = "0830e14cb03a2de2e7938ff96d4f74b6b47461a999fb866b6b1b13a66a5aa5b9",
)

load("@rules_multitool//multitool:multitool.bzl", "multitool")

multitool(
    name = "multitool",
    lockfile = "//tools:multitool.lock.json",
)
```
to this:
```bazel
bazel_dep(name = "rules_multitool", version = "0.12.0")

multitool = use_extension("@rules_multitool//multitool:extension.bzl", "multitool")
multitool.hub(lockfile = "//tools:multitool.lock.json")
use_repo(multitool, "multitool")

bazel_dep(name = "aspect_rules_lint", version = "v1.0.0-rc5")
git_override(
    module_name = "aspect_rules_lint",
    commit = "369cdb2725aa75a1de3cc580a051e8dca9b87c30",
    remote = "https://github.com/aspect-build/rules_lint",
)
```


When I reran tests, I got this:
```
INFO: Invocation ID: 72bcbc31-f6f7-4d37-ad7c-4e5e407c0ac2
INFO: Repository rules_multitool~~multitool~multitool.linux_x86_64 instantiated at:
  <builtin>: in <toplevel>
Repository rule _env_specific_tools defined at:
  $CACHEDIR/91970f834fe90649640cac0e23df1dee/external/rules_multitool~/multitool/private/multitool.bzl:224:38: in <toplevel>
INFO: repository @@rules_multitool~~multitool~multitool.linux_x86_64' used the following cache hits instead of downloading the corresponding file.
 * Hash 'd99b06506aee2ac9113daec3049922e70dc8cffb84658e3ae512c6a6cbe101b6' for https://$MIRROR_HOSTNAME_REDACTED/artifactory/github.com/mvdan/sh/releases/download/v3.9.0/shfmt_v3.9.0_linux_amd64
 * Hash 'a137c5e969609c3995c4d05817a247cfef8a92760c5306c3ad7df0355dd62970' for https://$MIRROR_HOSTNAME_REDACTED/artifactory/github.com/google/go-jsonnet/releases/download/v0.20.0/go-jsonnet_0.20.0_Linux_x86_64.tar.gz
 * Hash '6ff459c1dcae3b0b00844c1a5a4a5b0f547237d8a4f3624aaea8d424aeef24c6' for https://$MIRROR_HOSTNAME_REDACTED/artifactory/github.com/mvdan/gofumpt/releases/download/v0.7.0/gofumpt_v0.7.0_linux_amd64
 * Hash '3ff056b5e8259003f67fd0f0ed7229499cfb0b41f3ff55cc184088589994f7a5' for http://nonexistent.example.org/cacheprobe
If the definition of 'repository @@rules_multitool~~multitool~multitool.linux_x86_64' was updated, verify that the hashes were also updated.
ERROR: An error occurred during the fetch of repository 'rules_multitool~~multitool~multitool.linux_x86_64':
   Traceback (most recent call last):
        File "$CACHEDIR/91970f834fe90649640cac0e23df1dee/external/rules_multitool~/multitool/private/multitool.bzl", line 161, column 42, in _env_specific_tools_impl
                rctx.download_and_extract(
Error in download_and_extract: Expected a file with a .zip, .jar, .war, .aar, .tar, .tar.gz, .tgz, .tar.xz, .txz, .tar.zst, .tzst, .tar.bz2, .tbz, .ar or .deb suffix (got $CACHEDIR/91970f834fe90649640cac0e23df1dee/external/rules_multitool~~multitool~multitool.linux_x86_64/tools/terraform/linux_x86_64_archive/temp137024580245705393/cacheprobe)
ERROR: no such package '@@rules_multitool~~multitool~multitool.linux_x86_64//tools/ruff': Expected a file with a .zip, .jar, .war, .aar, .tar, .tar.gz, .tgz, .tar.xz, .txz, .tar.zst, .tzst, .tar.bz2, .tbz, .ar or .deb suffix (got $CACHEDIR/91970f834fe90649640cac0e23df1dee/external/rules_multitool~~multitool~multitool.linux_x86_64/tools/terraform/linux_x86_64_archive/temp137024580245705393/cacheprobe)
ERROR: $CACHEDIR/91970f834fe90649640cac0e23df1dee/external/rules_multitool~~multitool~multitool/toolchains/BUILD.bazel:17:24: @@rules_multitool~~multitool~multitool//toolchains:ruff_linux_x86_64_toolchain_info depends on @@rules_multitool~~multitool~multitool.linux_x86_64//tools/ruff:linux_x86_64_executable in repository @@rules_multitool~~multitool~multitool.linux_x86_64 which failed to fetch. no such package '@@rules_multitool~~multitool~multitool.linux_x86_64//tools/ruff': Expected a file with a .zip, .jar, .war, .aar, .tar, .tar.gz, .tgz, .tar.xz, .txz, .tar.zst, .tzst, .tar.bz2, .tbz, .ar or .deb suffix (got $CACHEDIR/91970f834fe90649640cac0e23df1dee/external/rules_multitool~~multitool~multitool.linux_x86_64/tools/terraform/linux_x86_64_archive/temp137024580245705393/cacheprobe)
ERROR: Analysis of target '//thirdparty/xxHash:pytest_ruff_lint' failed; build aborted: Analysis failed
```
So my initial analysis was wrong, it looks like the problem is that it's triggering the cacheprobe error for some reason. I think the reason I haven't been able to repro outside this repo is none of my repros have a prepopulated cache (but my attempts to repro in light of that knowledge so far have failed too, so...). `bazel clean --expunge` doesn't fix it; I suspect if I nuked my repo cache entirely it'd fix it, but then everyone else would have to nuke their caches, too, so I'm trying to avoid that!

I don't really know if this is multitool's fault, or a bazel core issue, or what, though, but hopefully you might know more? (I'm working on a repro for the second problem, stay tuned).

**Comment by novas0x2a** [comment]
Okay, reproed the windows hashing problem: https://github.com/novas0x2a/multitool-repro/tree/66343b183c888692b9227deccd28f697032fa902/windows

This `MODULE.bazel.lock` was generated on linux. When I switch to windows:
```
$ bazel build --lockfile_mode=error ...
ERROR: Analysis of target '//:version' failed; build aborted: MODULE.bazel.lock is no longer up-to-date because: The implementation of the extension 'ModuleExtensionId{bzlFileLabel=@@rules_multitool~//multitool:extension.bzl, extensionName=multitool, isolationKey=Optional.empty}' or one of its
 transitive .bzl files has changed. Please run `bazel mod deps --lockfile_mode=update` to update your lockfile.
INFO: Elapsed time: 0.289s, Critical Path: 0.00s
INFO: 1 process: 1 internal.
ERROR: Build did NOT complete successfully

$ bazel build ...
INFO: Analyzed target //:version (80 packages loaded, 387 targets configured).
INFO: Found 1 target...
Target //:version up-to-date:
  bazel-bin/version.txt
INFO: Elapsed time: 0.388s, Critical Path: 0.00s
INFO: 1 process: 1 internal.
INFO: Build completed successfully, 1 total action

$ git diff
diff --git a/windows/MODULE.bazel.lock b/windows/MODULE.bazel.lock
index 6cbade4..9af7ce6 100644
--- a/windows/MODULE.bazel.lock
+++ b/windows/MODULE.bazel.lock
@@ -1317,7 +1317,7 @@
     },
     "@@rules_multitool~//multitool:extension.bzl%multitool": {
       "general": {
-        "bzlTransitiveDigest": "G2CL37NxIhoaFlv/ma9cj0lSmLvPkIRHTKSLOFO7HFs=",
+        "bzlTransitiveDigest": "ZGQvjEv72CGwnaveOku2fDC1d17kq1Rx+RrJduWNKGU=",
         "usagesDigest": "j1TvvGZ+6Gsbx3NEhrCbsI0ivwEf8rHS91/kphF+Z5Y=",
         "recordedFileInputs": {},
         "recordedDirentsInputs": {},
```

**Comment by mark-thm** [comment]
In the repo you linked the committed MODULE.bazel.lock file does not seem to include any multitool details, which seems like the lock file was not in fact up to date?

**Comment by mark-thm** [comment]
For the problem you’re having with cacheprobe I think your multitool.lock.json file may have an error in it.

**Comment by novas0x2a** [comment]
> In the repo you linked the committed MODULE.bazel.lock file does not seem to include any multitool details, which seems like the lock file was not in fact up to date?

Hm, not sure what you mean, this is the field that is updated in the diff: https://github.com/novas0x2a/multitool-repro/blob/66343b183c888692b9227deccd28f697032fa902/windows/MODULE.bazel.lock#L1320

Also, the genrule runs fine, and doesn't change the `MODULE.bazel.lock`, and neither does `bazel mod deps --lockfile_mode=update`.

> For the problem you’re having with cacheprobe I think your multitool.lock.json file may have an error in it.

What sort of error? Note that the warnings that trigger the cacheprobe error are not from my multitool.lock.json, but the rules_lint one (if you're saying that the _aspect_ one has errors, maybe, I didn't validate that, but I'll take a look in a bit).

**Comment by mark-thm** [comment]
In particular this line:
> Error in download_and_extract: Expected a file with a .zip, .jar, .war, .aar, .tar, .tar.gz, .tgz, .tar.xz, .txz, .tar.zst, .tzst, .tar.bz2, .tbz, .ar or .deb suffix (got $CACHEDIR/91970f834fe90649640cac0e23df1dee/external/rules_multitool~~multitool~multitool.linux_x86_64/tools/terraform/linux_x86_64_archive/temp137024580245705393/cacheprobe)

suggests to me you may need to pass a type key (which will in turn be passed into download_and_extract)

**Comment by mark-thm** [comment]
I think it may be a Bazel bug in the MODULE.bazel.lock file computation — we don’t change the content of the bzl file depending on OS.

**Comment by mark-thm** [comment]
Sorry, I misread the MODULE.bazel.lock in my first pass -- I still think there's possibly a Bazel bug here, but want to check, is there additional diff content besides what you've shared here?

**Comment by novas0x2a** [comment]
Re: the cacheprobe, so, that's not a multitool-pinned file, that's https://github.com/bazelbuild/bazel/blob/225d0a8b4e16caefbba8841603b988d515036c65/src/main/java/com/google/devtools/build/lib/bazel/repository/downloader/DownloadManager.java#L220

I don't know what's going on here, though. My assumption is that those warnings were triggered by something that also triggered this cacheprobe failure.

**Comment by mark-thm** [comment]
Perhaps your terraform tooling is missing a type then? Absent that, again, this seems like a Bazel bug.

**Comment by novas0x2a** [comment]
@alexeagle figured out that https://github.com/theoremlp/rules_multitool/issues/59 is at least part of what's triggering the `cacheprobe` issue, and might actually be the entire issue (we don't have bazeldownloader redirects for every hostname in the fused multitool hub, just for the ones we actually use, so it's likely triggering the cacheprobe because it's downloading packages we don't use).

**Comment by alexeagle** [comment]
Great, maybe you could close this issue then? And if we're still asking rules_multitool maintainers to look at any other issue, give them a smaller more focused issue for that?

**Comment by mark-thm** [comment]
Seems like this should be fixed by #62, released as 0.15.0
